src/sentry/migrations/0866_backfill_open_periods.py

```diff
@@ -0,0 +1,177 @@
+# Generated by Django 5.1.7 on 2025-04-09 23:57
+
+import logging
+from datetime import datetime
+from enum import Enum
+from typing import TYPE_CHECKING, Any
+
+from django.db import migrations
+from django.db.backends.base.schema import BaseDatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+from django.db.models import Q
+
+from sentry.new_migrations.migrations import CheckedMigration
+from sentry.utils.query import RangeQuerySetWrapperWithProgressBarApprox
+
+if TYPE_CHECKING:
+    from sentry.models.activity import Activity as ActivityModelType
+
+
+logger = logging.getLogger(__name__)
+
+
+# copied constants and enums
+class ActivityType(Enum):
+    SET_RESOLVED = 1
+    SET_UNRESOLVED = 2
+    SET_REGRESSION = 6
+
+
+class GroupStatus:
+    UNRESOLVED = 0
+    RESOLVED = 1
+
+
+# end copy
+
+BATCH_SIZE = 100
+
+
+def get_open_periods_for_group(
+    apps: StateApps,
+    group_id: int,
+    status: int,
+    project_id: int,
+    first_seen: datetime,
+    query_end: datetime | None,
+    GroupOpenPeriod: Any,
+) -> list[Any]:
+    Activity = apps.get_model("sentry", "Activity")
+
+    # Filter to REGRESSION and RESOLVED activties to find the bounds of each open period.
+    # The only UNRESOLVED activity we would care about is the first UNRESOLVED activity for the group creation,
+    # but we don't create an entry for that.
+    filters = [
+        Q(group_id=group_id),
+        Q(type__in=[ActivityType.SET_REGRESSION.value, ActivityType.SET_RESOLVED.value]),
+    ]
+    if query_end:
+        filters.append(Q(datetime__lte=query_end))
+    activities = Activity.objects.filter(*filters).order_by("-datetime")
+    open_periods = []
+    start: datetime | None = None
+    end: datetime | None = None
+    end_activity: ActivityModelType | None = None
+
+    # Handle currently open period
+    if status == GroupStatus.UNRESOLVED and len(activities) > 0:
+        open_periods.append(
+            GroupOpenPeriod(
+                group_id=group_id,
+                project_id=project_id,
+                date_started=activities[0].datetime,
+                date_ended=None,
+                resolution_activity=None,
+                user_id=None,
+            )
+        )
+        activities = activities[1:]
+
+    for activity in activities:
+        if activity.type == ActivityType.SET_RESOLVED.value:
+            end = activity.datetime
+            end_activity = activity
+        elif activity.type == ActivityType.SET_REGRESSION.value:
+            start = activity.datetime
+            if start is not None and end is not None:
```

Comment:
Worth logging an error is `end is None`? It probably shouldn't happen, I think?

Reply:
oops found an example: https://redash.getsentry.net/queries/7532 (SENTRY-3RYH)